src/validator.cpp

```diff
@@ -1635,6 +1708,530 @@ void Validator::ValidatorImpl::validateMathMLElements(const XmlNodePtr &node, co
+        }
+    } else if (node->isMathmlElement("piece")) {
+        if (hasTwoMathmlChildren(node, component)) {
+            for (size_t i = 0, iMax = mathmlChildCount(node); i < iMax; ++i) {
+                validateMathMLElementsChildrenAndSiblings(mathmlChildNode(node, i), component);
+            }
+        }
+    } else if (node->isMathmlElement("otherwise")) {
+        if (hasOneMathmlChild(node, component)) {
+            for (size_t i = 0, iMax = mathmlChildCount(node); i < iMax; ++i) {
+                validateMathMLElementsChildrenAndSiblings(mathmlChildNode(node, i), component);
+            }
+        }
+
+        // Token elements.
+
+    } else if (node->isMathmlElement("ci")) {
+        auto ok = (nonCommentChildCount(node) != 1) ? false : !nonCommentChildNode(node, 0)->convertToStrippedString().empty();
+
+        if (!ok) {
+            addMathMLIssue("Math has a 'ci' element with no identifier as a child.",
+                           Issue::ReferenceRule::MATH_CI_VARIABLE_REF,
+                           component);
+        }
+    } else if (node->isMathmlElement("cn")) {
+        auto cnBase = node->attribute("base");
+
+        if (!cnBase.empty() && (cnBase != "10")) {
+            addMathMLIssue("Math has a 'cn' element which is not in base 10.",
+                           Issue::ReferenceRule::MATH_CN_BASE10,
+                           component);
+
+            return;
+        }
+
+        auto cnType = node->attribute("type");
+
+        if (cnType.empty() || (cnType == "real")) {
+            auto ok = (nonCommentChildCount(node) != 1) ? false : nonCommentChildNode(node, 0)->isBasicReal();
+
+            if (!ok) {
+                addMathMLIssue("Math has a 'cn' element of 'real' type with no valid text node (representing a basic number) as a child.",
+                               Issue::ReferenceRule::MATH_CN_FORMAT,
+                               component);
+            }
+        } else if (cnType == "e-notation") {
+            auto ok = false;
+
+            if (nonCommentChildCount(node) == 3) {
+                ok = nonCommentChildNode(node, 0)->isBasicReal()
+                     && nonCommentChildNode(node, 1)->isMathmlElement("sep")
+                     && nonCommentChildNode(node, 2)->isInteger();
+            }
+
+            if (!ok) {
+                addMathMLIssue("Math has a 'cn' element of 'e-notation' type with no valid text node (representing a basic number), no 'sep' element, and/or no valid text node (representing an integer) as children.",
+                               Issue::ReferenceRule::MATH_CN_FORMAT,
+                               component);
+            }
+        } else {
+            addMathMLIssue("Math has a 'cn' element which is not of 'real' or 'e-notation' type.",
+                           Issue::ReferenceRule::MATH_CN_FORMAT,
+                           component);
+        }
+
+        // Qualifier elements.
+
+    } else if (node->isMathmlElement("degree")) {
```

Comment:
this check is duplicating the checks being done for `root` and `bvar`. Perhaps we can skip these checks?

Reply:
No, it doesn't. Someone could perfectly use `degree` without using `root` or `bvar`, and this is what those checks are about. See [this CellML file](https://github.com/agarny/libcellml/blob/issue1082/tests/resources/invalidmathmlelementschildrenorsiblings.cellml) and search for `degree` to see what I mean.